matrix-room.c

```diff
@@ -503,6 +520,142 @@ static void _send_image_hook(MatrixRoomEvent *event, gboolean just_free)
     }
 }
 
+/* Passed through matrix_api_download_file all the way
+ * downto _image_download_complete
+ */
+struct ReceiveImageData {
+    PurpleConversation *conv;
+    const gchar *content_uri;
+    gint64 timestamp;
+    const gchar *room_id;
+    const gchar *sender_display_name;
+};
+
+static void _image_download_complete(MatrixConnectionData *ma,
+          gpointer user_data, JsonNode *json_root,
+          const char *raw_body, size_t raw_body_len, const char *content_type)
+{
+    struct ReceiveImageData *rid = user_data;
+    if (check_image_type(content_type)) {
+        /* Excellent - something to work with */
+        int img_id = purple_imgstore_add_with_id(g_memdup(raw_body, raw_body_len),
+                                                 raw_body_len, NULL);
+        serv_got_chat_in(rid->conv->account->gc, g_str_hash(rid->room_id), rid->sender_display_name,
+                PURPLE_MESSAGE_RECV | PURPLE_MESSAGE_IMAGES,
+                g_strdup_printf("<IMG ID=\"%d\">", img_id), rid->timestamp / 1000);
+    } else {
+        serv_got_chat_in(rid->conv->account->gc, g_str_hash(rid->room_id),
```

Comment:
it would be nice to show the fallback (event.body) here?

ditto in _image_download_bad_response and _image_download_error